File: week1_baseline/ruby/10_standard_tool_library/test/test_config_mcp_servers.rb
require "minitest/autorun"
require "tmpdir"
require "yaml"

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)

require "boukensha"

# Config#mcp_servers — servers as data, not code.
class TestConfigMcpServers < Minitest::Test
  def with_settings(hash)
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, "settings.yaml"), YAML.dump(hash))
      previous = ENV["BOUKENSHA_DIR"]
      ENV["BOUKENSHA_DIR"] = dir
      begin
        yield Boukensha::Config.new
      ensure
        ENV["BOUKENSHA_DIR"] = previous
      end
    end
  end

  def test_absent_block_is_empty_not_nil
    with_settings({ "tasks" => {} }) do |cfg|
      assert_equal({}, cfg.mcp_servers)
    end
  end

  def test_parses_a_full_entry
    with_settings({
      "mcp_servers" => {
        "mud" => {
          "command" => "mud-manager",
          "args" => ["--mcp"],
          "prefix" => "tbamud",
          "env" => { "MUD_HOST" => "localhost", "MUD_PORT" => 4000 }
        }
      }
    }) do |cfg|
      entry = cfg.mcp_servers["mud"]

      assert_equal "mud-manager", entry[:command]
      assert_equal ["--mcp"], entry[:args]
      assert_equal "tbamud", entry[:prefix]
      # Env values are stringified — a YAML integer port would otherwise reach
      # Open3 as an Integer and blow up on spawn.
      assert_equal({ "MUD_HOST" => "localhost", "MUD_PORT" => "4000" }, entry[:env])
    end
  end

  def test_defaults
    with_settings({ "mcp_servers" => { "bare" => { "command" => "thing" } } }) do |cfg|
      entry = cfg.mcp_servers["bare"]

      assert_equal [], entry[:args]
      assert_equal({}, entry[:env])
      assert_nil entry[:prefix]
      # A server you bothered to configure and which then fails is a problem you
      # want to hear about, so required defaults to true.
      assert_equal true, entry[:required]
    end
  end

  def test_required_can_be_turned_off
    with_settings({
      "mcp_servers" => { "decor" => { "command" => "thing", "required" => false } }
    }) do |cfg|
      assert_equal false, cfg.mcp_servers["decor"][:required]
    end
  end

  def test_malformed_entries_do_not_raise
    # A half-written config should surface as "no command" at registration, not
    # as a parse crash before the agent even starts.
    with_settings({ "mcp_servers" => { "broken" => nil, "alsobroken" => "nonsense" } }) do |cfg|
      assert_nil cfg.mcp_servers["broken"][:command]
      assert_nil cfg.mcp_servers["alsobroken"][:command]
    end
  end

  def test_multiple_servers
    with_settings({
      "mcp_servers" => {
        "mud" => { "command" => "mud-manager", "args" => ["--mcp"] },
        "filesystem" => { "command" => "npx", "args" => ["-y", "server-filesystem", "/tmp"] }
      }
    }) do |cfg|
      assert_equal %w[filesystem mud], cfg.mcp_servers.keys.sort
    end
  end
end

File: week1_baseline/python/11_tui/test/test_config_mcp_servers.py
import os
import sys
import tempfile
import unittest
from contextlib import contextmanager
from pathlib import Path

import yaml

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import boukensha  # noqa: E402


@contextmanager
def settings(hash_):
    """Point BOUKENSHA_DIR at a throwaway directory holding one settings.yaml."""
    with tempfile.TemporaryDirectory() as dir_:
        Path(dir_, "settings.yaml").write_text(yaml.dump(hash_))
        previous = os.environ.get("BOUKENSHA_DIR")
        os.environ["BOUKENSHA_DIR"] = dir_
        try:
            yield boukensha.Config()
        finally:
            if previous is None:
                os.environ.pop("BOUKENSHA_DIR", None)
            else:
                os.environ["BOUKENSHA_DIR"] = previous


class TestConfigMcpServers(unittest.TestCase):
    """mcp_servers — servers as data, not code."""

    def test_absent_block_is_empty_not_none(self):
        with settings({"tasks": {}}) as cfg:
            self.assertEqual({}, cfg.mcp_servers)

    def test_parses_a_full_entry(self):
        with settings(
            {
                "mcp_servers": {
                    "mud": {
                        "command": "mud-manager",
                        "args": ["--mcp"],
                        "prefix": "tbamud",
                        "env": {"MUD_HOST": "localhost", "MUD_PORT": 4000},
                    }
                }
            }
        ) as cfg:
            entry = cfg.mcp_servers["mud"]

            self.assertEqual("mud-manager", entry["command"])
            self.assertEqual(["--mcp"], entry["args"])
            self.assertEqual("tbamud", entry["prefix"])
            # Env values are stringified — a YAML integer port would otherwise
            # reach the subprocess environment as an int and raise on spawn.
            self.assertEqual({"MUD_HOST": "localhost", "MUD_PORT": "4000"}, entry["env"])

    def test_defaults(self):
        with settings({"mcp_servers": {"bare": {"command": "thing"}}}) as cfg:
            entry = cfg.mcp_servers["bare"]

            self.assertEqual([], entry["args"])
            self.assertEqual({}, entry["env"])
            self.assertIsNone(entry["prefix"])
            # A server you bothered to configure and which then fails is a
            # problem you want to hear about, so required defaults to True.
            self.assertTrue(entry["required"])

    def test_required_can_be_turned_off(self):
        with settings(
            {"mcp_servers": {"decor": {"command": "thing", "required": False}}}
        ) as cfg:
            self.assertFalse(cfg.mcp_servers["decor"]["required"])

    def test_malformed_entries_do_not_raise(self):
        # A half-written config should surface as "no command" at registration,
        # not as a parse crash before the agent even starts.
        with settings({"mcp_servers": {"broken": None, "alsobroken": "nonsense"}}) as cfg:
            self.assertIsNone(cfg.mcp_servers["broken"]["command"])
            self.assertIsNone(cfg.mcp_servers["alsobroken"]["command"])

    def test_multiple_servers(self):
        with settings(
            {
                "mcp_servers": {
                    "mud": {"command": "mud-manager", "args": ["--mcp"]},
                    "filesystem": {"command": "npx", "args": ["-y", "server-fs", "/tmp"]},
                }
            }
        ) as cfg:
            self.assertEqual(["filesystem", "mud"], sorted(cfg.mcp_servers))


class TestMcpOptionTriState(unittest.TestCase):
    """None, False and True are three different answers.

    Ruby distinguishes nil / false / Hash on this option; in Python None and
    False are both falsy, so a naive `if not mcp` would collapse the first two.
    That bug is invisible in normal use — it only shows for someone explicitly
    disabling the route, which is exactly the person least likely to be running
    the tests. Hence this test.
    """

    def test_none_and_false_both_register_nothing(self):
        from boukensha.run import _mcp_opts

        with settings({"mud": {"host": "h", "username": "u", "password": "p"}}) as cfg:
            self.assertIsNone(_mcp_opts(None, cfg))
            self.assertIsNone(_mcp_opts(False, cfg))

    def test_true_builds_the_mud_preset(self):
        from boukensha.run import _mcp_opts

        with settings({"mud": {"host": "h", "username": "u", "password": "p"}}) as cfg:
            opts = _mcp_opts(True, cfg)

            self.assertEqual(["--mcp"], opts["args"])
            self.assertEqual(boukensha.MUD_PREFIX, opts["prefix"])
            self.assertEqual("u", opts["env"]["MUD_NAME"])

    def test_config_entry_wins_over_the_preset(self):
        from boukensha.run import _mcp_opts

        with settings(
            {
                "mud": {"host": "h", "username": "u", "password": "p"},
                "mcp_servers": {"mud": {"command": "/custom/mud-manager", "prefix": "zzz"}},
            }
        ) as cfg:
            opts = _mcp_opts(True, cfg)

            self.assertEqual("/custom/mud-manager", opts["command"])
            self.assertEqual("zzz", opts["prefix"])
            # Credentials still layer in from the mud: block.
            self.assertEqual("u", opts["env"]["MUD_NAME"])

    def test_dict_overrides_parts_of_the_preset(self):
        from boukensha.run import _mcp_opts

        with settings({"mud": {"host": "h", "username": "u", "password": "p"}}) as cfg:
            opts = _mcp_opts({"prefix": "custom"}, cfg)

            self.assertEqual("custom", opts["prefix"])
            self.assertEqual(["--mcp"], opts["args"])


if __name__ == "__main__":
    unittest.main()
